internal/api/mfa.go

```diff
@@ -196,6 +200,56 @@ func (a *API) ChallengeFactor(w http.ResponseWriter, r *http.Request) error {
 	})
 }
 
+func (a *API) invokeHook(ctx context.Context, input any, output any) *hooks.AuthHookError {
+	var response []byte
+	switch input.(type) {
+	case hooks.MFAVerificationAttemptInput:
+		payload, err := json.Marshal(&input)
+		if err != nil {
+			panic(err)
+		}
+		hookName, err := hooks.FetchHookName(a.config.Hook.MFAVerificationAttempt)
+		if err != nil {
+			return &hooks.AuthHookError{
+				Message: "invalid hook name",
+			}
+		}
+		if err := a.db.Transaction(func(tx *storage.Connection) error {
+			// We rely on Postgres timeouts to ensure the function doesn't overrun
+			timeoutQuery := tx.RawQuery(fmt.Sprintf("set local statement_timeout TO '%d';", hooks.DefaultTimeout))
+			if terr := timeoutQuery.Exec(); terr != nil {
+				return terr
+			}
+			query := tx.RawQuery(fmt.Sprintf("SELECT * from %s(?)", hookName), payload)
+			terr := query.First(&response)
+			if terr != nil {
+				return terr
+			}
+			return nil
+		}); err != nil {
+			return &hooks.AuthHookError{
+				Message: err.Error(),
+			}
+		}
+		// As we the response fields aren't known to us we try to check if it's an error first.
+		hookResponseOrError := hooks.AuthHookErrorResponse{}
+		err = json.Unmarshal(response, &hookResponseOrError)
+		if err == nil && hookResponseOrError.IsError() {
+			return &hookResponseOrError.AuthHookError
+		}
+		if err = json.Unmarshal(response, output); err != nil {
+			return &hooks.AuthHookError{
+				Message: "error unmarshalling response",
+			}
+		}
```

Comment:
Why do this double unmarshal? Can we use a single "Response" or "Output" type instead that encodes both the error and success?

If it contains a non-zero error value, then you can return an error _in addition to_ unmarshalling into `output`?